### PR TITLE
ILE-9 SMS MFA

### PR DESCRIPTION
**Pull request by dana-kainos**

### CODE REVIEW

File: views/auth/mfa.njk [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    } if errormessage else false,
+    label: {
+        text: "OTP code"
+    }
+  }) }}
+
+    {{ govukButton({
+        text: "Submit",
+        preventDoubleClick: true
+      }) }}
```
Comment (declanlarkin) [inline views/auth/mfa.njk:36]: indentation

File: views/auth/mfa.njk [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+{% from "govuk/components/input/macro.njk" import govukInput %}
+{% from "govuk/components/password-input/macro.njk" import govukPasswordInput %}
+{% from "govuk/components/notification-banner/macro.njk" import govukNotificationBanner %}
+
+{% block content %}
+
+<main class="govuk-main-wrapper">
+    <div class="govuk-grid-row">
+
+<div class="govuk-grid-column-two-thirds">
```
Comment (declanlarkin) [inline views/auth/mfa.njk]: Clean up file indentation

File: views/auth/mfa.njk [hunk 3]
```diff
@@ -0,0 +1,46 @@
+{% extends "templates/layout.njk" %}
+
+{% from "govuk/components/input/macro.njk" import govukInput %}
+{% from "govuk/components/password-input/macro.njk" import govukPasswordInput %}
+{% from "govuk/components/notification-banner/macro.njk" import govukNotificationBanner %}
+
```
Comment (declanlarkin) [inline views/auth/mfa.njk:4]: govukPasswordInput & govukNotificationBanner not used?

File: src/controllers/AuthController.ts [hunk 4]
```diff
@@ -1,14 +1,32 @@
...
+import { JwtToken } from "../models/JwtToken";
+import { jwtDecode } from "jwt-decode";
 
 export const getLoginForm = async (req: express.Request, res: express.Response): Promise<void> => {
     res.render('auth/login.njk');
 }
 
 export const postLoginForm = async (req: express.Request, res: express.Response): Promise<void> => {
     try {
+        if (req.body.mfaCode) {
```
Comment (declanlarkin) [inline src/controllers/AuthController.ts:12]: Just to confirm this postLoginForm is called on inital log in and also called when logging in with MFA (from otp entry screen)?

File: src/controllers/AuthController.ts [hunk 5]
```diff
@@ -1,14 +1,32 @@
...
+                return res.render('auth/mfa.njk', { errormessage: "Invalid OTP code." });
+            }
+        }
+
         req.session.token = await getToken(req.body);
-        res.redirect('/sales-employees');
+        
+        const decodedToken: JwtToken = jwtDecode(req.session.token);
+
+        if (!decodedToken.User.phone) return res.redirect('/sales-employees');
```
Comment (declanlarkin) [inline src/controllers/AuthController.ts:26]: Assume phone number is optional when creating an employee?

Comment (dana-kainos) [reply src/controllers/AuthController.ts:26]: yes, mfa will only be requested if the user has a phone number assigned to them on creation

File: src/controllers/AuthController.ts (hunk 4 above)
Comment (dana-kainos) [reply src/controllers/AuthController.ts:12]: yep

**Review by declanlarkin — APPROVED** [review]

**Review by lukemKainos — APPROVED** [review]

**Review by dylanmckk — APPROVED** [review]

**Pull request merged**